apps/aecore/src/aec_conductor.erl

```diff
@@ -0,0 +1,601 @@
+        aec_chain:get_transactions_between(Hash1, Ancestor),
+    {ok, TransactionsOnNewChain} =
+        aec_chain:get_transactions_between(Hash2, Ancestor),
+    ok = aec_tx_pool:fork_update(TransactionsOnNewChain, TransactionsOnOldChain),
+    ok.
+
+%%%===================================================================
+%%% Worker: Wait for keys to appear
+
+-define(WAIT_FOR_KEYS_RETRIES, 10).
+
+wait_for_keys(State) ->
+    Fun = fun wait_for_keys_worker/0,
+    dispatch_worker(wait_for_keys, Fun, State).
+
+wait_for_keys_worker() ->
+    wait_for_keys_worker(?WAIT_FOR_KEYS_RETRIES).
+
+wait_for_keys_worker(0) ->
+    timeout;
+wait_for_keys_worker(N) ->
+    case aec_keys:pubkey() of
+        {ok, _Pubkey} -> keys_ready; %% TODO: We could keep the key in the state
+        {error, _} ->
+            timer:sleep(500),
+            wait_for_keys_worker(N - 1)
+    end.
+
+handle_wait_for_keys_reply(keys_ready, State) ->
+    create_block_candidate(State);
+handle_wait_for_keys_reply(timeout, State) ->
+    %% TODO: We should probably die hard at some point instead of retrying.
+    epoch_mining:error("Timed out waiting for keys. Retrying."),
+    wait_for_keys(State).
+
+%%%===================================================================
+%%% Worker: Start mining
+
+start_mining(#state{mining_state = 'stopped'} = State) ->
+    State;
+start_mining(#state{block_candidate = undefined} = State) ->
+    %% We need to generate a new block candidate first.
+    create_block_candidate(State);
+start_mining(#state{block_candidate = #candidate{top_hash = OldHash},
+                    seen_top_block_hash = SeenHash
+                   } = State) when OldHash =/= SeenHash ->
+    %% Candidate generated with stale top hash.
+    %% Regenerate the candidate.
+    create_block_candidate(State#state{block_candidate = undefined});
+start_mining(#state{} = State) ->
+    epoch_mining:info("Starting mining"),
+    #candidate{block = Block,
+               nonce = Nonce} = State#state.block_candidate,
+    Info = [{top_block_hash, State#state.seen_top_block_hash}],
+    aec_events:publish(start_mining, Info),
+    Fun = fun() -> {aec_mining:mine(Block, Nonce),
+                    State#state.block_candidate}
+          end,
+    dispatch_worker(mining, Fun, State).
+
+handle_mining_reply({{ok, Block},_Candidate}, State) ->
+    %% TODO: This should listen on some event instead
+    ws_handler:broadcast(miner, mined_block,
+                         [{height, aec_blocks:height(Block)}]),
+    epoch_mining:info("Miner ~p finished with ~p ~n", [self(), {ok, Block}]),
+    State1 = save_mined_block(Block, State),
+    State2 = State1#state{block_candidate = undefined},
+    case preempt_if_new_top(State2) of
+        no_change         -> start_mining(State2);
+        {changed, State3} -> start_mining(State3)
+    end;
+handle_mining_reply({{error, no_solution}, Candidate}, State) ->
+    try exometer:update([ae,epoch,aecore,mining,retries], 1)
```

Comment:
I have in mind to add an `aec_metrics:try_update/2` function, since this is a very common pattern.